File: core-go/engine_test.go
package downloadcore

import (
	"bytes"
	"context"
	"errors"
	"os"
	"path/filepath"
	"testing"
	"time"

	"downloadcore/testserver"
)

func testConfig(t *testing.T) Config {
	t.Helper()
	c := DefaultConfig()
	c.InitialThreads = 1
	c.MaxThreads = 4
	c.MinPartSize = 256 << 10
	c.BufferSize = 64 << 10
	c.IdleTimeout = 2 * time.Second
	c.MaxRetries = 2
	c.RetryDelay = 20 * time.Millisecond
	c.TempDir = t.TempDir()
	return c
}

func makeData(n int, seed byte) []byte {
	b := make([]byte, n)
	for i := range b {
		b[i] = byte((i*31 + int(seed)*7) & 0xff)
	}
	return b
}

func requireFileEquals(t *testing.T, path string, want []byte) {
	t.Helper()
	got, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("读取文件失败: %v", err)
	}
	if !bytes.Equal(got, want) {
		t.Fatalf("文件内容不一致: got %d 字节, want %d 字节", len(got), len(want))
	}
}

// --- 单元测试：切分与分裂 ---

func TestSplitToRange(t *testing.T) {
	r := splitToRange(10, 3, 4)
	if len(r) != 4 {
		t.Fatalf("期望 4 段，实际 %d", len(r))
	}
	if r[0][0] != 0 || r[len(r)-1][1] != 9 {
		t.Fatalf("段未覆盖整个文件: %v", r)
	}
	var sum int64
	for i, s := range r {
		sum += s[1] - s[0] + 1
		if i > 0 && s[0] != r[i-1][1]+1 {
			t.Fatalf("段之间不连续: %v", r)
		}
	}
	if sum != 10 {
		t.Fatalf("总长度应为 10，实际 %d", sum)
	}
}

func TestPartSplitKeepsPrefixAndSafeZone(t *testing.T) {
	origTo := int64(10)*safetyStep - 1
	p := newPart(0, origTo)
	sz := p.snapshot
	_, _, _, safeBefore := sz()
	np := p.split()
	if np == nil {
		t.Fatal("应当可以分裂")
	}
	if np.from != p.to+1 {
		t.Fatalf("两段必须首尾相接: old.to=%d new.from=%d", p.to, np.from)
	}
	if np.to != origTo {
		t.Fatalf("后半段应当接到原末尾: %d != %d", np.to, origTo)
	}
	if p.to < safeBefore {
		t.Fatalf("分裂后前半段不能小于安全区: to=%d safe=%d", p.to, safeBefore)
	}
}

// 分裂必须优先切"剩下活最多"的那段，否则负载会严重不均。
func TestSplitPicksLargestRemaining(t *testing.T) {
	j := &job{cfg: Config{MinPartSize: safetyStep}}
	a := newPart(0, 1<<20-1)     // 1 MiB，太小，不该被切
	b := newPart(1<<20, 9<<20-1) // 8 MiB，应当被切
	j.parts = []*part{a, b}

	np := j.splitOneLocked()
	if np == nil {
		t.Fatal("应当能分裂")
	}
	if np.from <= 1<<20 {
		t.Fatalf("应当拆分更大的那段，实际新增 from=%d", np.from)
	}
	if a.to != 1<<20-1 {
		t.Fatalf("小段不该被动：to=%d", a.to)
	}
}

// --- 端到端测试 ---

func TestDownloadDynamicSegmentation(t *testing.T) {
	data := makeData(4<<20, 1)
	srv := testserver.New(data)
	defer srv.Close()

	e := New(testConfig(t))
	target := filepath.Join(t.TempDir(), "out.bin")

	res, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{})
	if err != nil {
		t.Fatalf("下载失败: %v", err)
	}
	if !res.RangeOK {
		t.Fatal("应当支持分段")
	}
	if res.Parts < 2 {
		t.Fatalf("期望动态分段产生多于 1 段，实际 %d", res.Parts)
	}
	requireFileEquals(t, target, data)
}

func TestNoRangeFallsBackToSingleStream(t *testing.T) {
	data := makeData(1<<20, 2)
	srv := testserver.New(data)
	defer srv.Close()
	srv.SetNoRange(true)

	e := New(testConfig(t))
	target := filepath.Join(t.TempDir(), "out.bin")

	res, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{})
	if err != nil {
		t.Fatalf("下载失败: %v", err)
	}
	if res.RangeOK {
		t.Fatal("服务器不支持分段，RangeOK 应为 false")
	}
	requireFileEquals(t, target, data)
}

func TestRangeMismatchDetected(t *testing.T) {
	data := makeData(1<<20, 3)
	srv := testserver.New(data)
	defer srv.Close()
	srv.SetFailRange(true)

	cfg := testConfig(t)
	cfg.InitialThreads = 1
	cfg.MaxThreads = 1
	cfg.MaxRetries = 1
	cfg.RetryDelay = 10 * time.Millisecond
	e := New(cfg)
	target := filepath.Join(t.TempDir(), "out.bin")

	_, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{})
	if err == nil {
		t.Fatal("服务器返回错误分段，应当报错")
	}
	if !errors.Is(err, ErrTooManyFailures) {
		t.Fatalf("期望重试耗尽错误，实际: %v", err)
	}
}

func TestIdleTimeoutRetries(t *testing.T) {
	data := makeData(2<<20, 4)
	srv := testserver.New(data)
	defer srv.Close()
	srv.SetStallAfter(64 << 10)

	cfg := testConfig(t)
	cfg.InitialThreads = 1
	cfg.MaxThreads = 1
	cfg.IdleTimeout = 300 * time.Millisecond
	cfg.MaxRetries = 1
	cfg.RetryDelay = 10 * time.Millisecond
	e := New(cfg)
	target := filepath.Join(t.TempDir(), "out.bin")

	_, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{})
	if err == nil {
		t.Fatal("连接卡住，最终应当报错")
	}
	if !errors.Is(err, ErrTooManyFailures) {
		t.Fatalf("期望重试耗尽错误，实际: %v", err)
	}
}

func TestCancelThenResume(t *testing.T) {
	data := makeData(4<<20, 5)
	srv := testserver.New(data)
	defer srv.Close()
	srv.SetSpeed(2 << 20) // 2 MiB/s，保证能中途取消

	cfg := testConfig(t)
	cfg.MaxRetries = 0
	cfg.IdleTimeout = 5 * time.Second
	e := New(cfg)
	target := filepath.Join(t.TempDir(), "out.bin")

	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		time.Sleep(400 * time.Millisecond)
		cancel()
	}()

	_, err := e.Download(ctx, Request{URL: srv.URL(), TargetFile: target}, Callbacks{})
	if !errors.Is(err, context.Canceled) {
		t.Fatalf("期望取消错误，实际: %v", err)
	}
	if _, statErr := os.Stat(target); !os.IsNotExist(statErr) {
		t.Fatal("取消后不应存在最终文件")
	}

	// 恢复：全速重下剩余部分
	srv.SetSpeed(0)
	res, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{})
	if err != nil {
		t.Fatalf("续传失败: %v", err)
	}
	requireFileEquals(t, target, data)
	_ = res
}

func TestFileChangedOnResumeStartsFresh(t *testing.T) {
	data1 := makeData(2<<20, 6)
	data2 := makeData(2<<20, 7)
	srv := testserver.New(data1)
	defer srv.Close()
	srv.SetSpeed(1 << 20)

	cfg := testConfig(t)
	cfg.MaxRetries = 0
	cfg.IdleTimeout = 5 * time.Second
	e := New(cfg)
	target := filepath.Join(t.TempDir(), "out.bin")

	ctx, cancel := context.WithCancel(context.Background())
	go func() {
		time.Sleep(300 * time.Millisecond)
		cancel()
	}()
	if _, err := e.Download(ctx, Request{URL: srv.URL(), TargetFile: target}, Callbacks{}); !errors.Is(err, context.Canceled) {
		t.Fatalf("期望取消错误，实际: %v", err)
	}

	// 服务器上的文件变了
	srv.SetData(data2, `"v2"`)
	srv.SetSpeed(0)

	if _, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{}); err != nil {
		t.Fatalf("重新下载失败: %v", err)
	}
	// 必须是"新文件"，绝不能新旧内容拼在一起
	requireFileEquals(t, target, data2)
}

// 某条连接被限速到爬行时，看门狗应当把它重开，让下载恢复正常速度。
func TestSlowConnectionGetsReconnected(t *testing.T) {
	data := makeData(1<<20, 9)
	srv := testserver.New(data)
	defer srv.Close()
	// 从偏移 0 开始的请求只有 64 KB/s（模拟"某条连接被限速"）；
	// 从别的偏移开始是全速 —— 所以重开连接后应当立刻变快。
	srv.SetSlowAtZero(64 << 10)

	oldW, oldB, oldR := slowWindow, slowMinBytes, slowRemainingMin
	slowWindow, slowMinBytes, slowRemainingMin = 200*time.Millisecond, 64<<10, 0
	defer func() { slowWindow, slowMinBytes, slowRemainingMin = oldW, oldB, oldR }()

	cfg := testConfig(t)
	cfg.InitialThreads = 1
	cfg.MaxThreads = 1
	e := New(cfg)
	target := filepath.Join(t.TempDir(), "out.bin")

	start := time.Now()
	if _, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetFile: target}, Callbacks{}); err != nil {
		t.Fatalf("下载失败: %v", err)
	}
	elapsed := time.Since(start)
	requireFileEquals(t, target, data)
	// 没有看门狗的话，1MB 以 64KB/s 下要 ~16 秒；这里应当在几秒内完成。
	if elapsed > 5*time.Second {
		t.Fatalf("慢连接没有被重开，用时 %v", elapsed)
	}
}

func TestAutoFileNameWhenTargetMissing(t *testing.T) {
	data := makeData(512<<10, 8)
	srv := testserver.New(data)
	defer srv.Close()

	e := New(testConfig(t))
	dir := t.TempDir()

	// 只给网址，不给文件名 —— 引擎应自己取名
	res, err := e.Download(context.Background(), Request{URL: srv.URL(), TargetDir: dir}, Callbacks{})
	if err != nil {
		t.Fatalf("下载失败: %v", err)
	}
	if filepath.Dir(res.Path) != dir {
		t.Fatalf("文件没有落到指定目录: %s", res.Path)
	}
	requireFileEquals(t, res.Path, data)
}

File: core-rs/tests/engine_test.rs
//! 端到端测试：用一个 std 实现的、可摆布的测试 HTTP 服务器，
//! 验证多线程、动态分段、不支持分段、对暗号、空闲超时、取消续传、换文件。

use downloadcore::{Callbacks, Config, Engine, ErrorKind, Request};
use std::io::{BufRead, BufReader, Write};
use std::net::{SocketAddr, TcpListener, TcpStream};
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, AtomicI64, AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

// ---------------- 测试服务器 ----------------

struct TestServer {
    addr: SocketAddr,
    data: Arc<Mutex<Vec<u8>>>,
    etag: Arc<Mutex<String>>,
    no_range: Arc<AtomicBool>,
    fail_range: Arc<AtomicBool>,
    stall_after: Arc<AtomicI64>,
    speed: Arc<AtomicI64>,
    hits: Arc<AtomicUsize>,
    stop: Arc<AtomicBool>,
    handle: Option<thread::JoinHandle<()>>,
}

impl TestServer {
    fn new(data: Vec<u8>) -> Self {
        let listener = TcpListener::bind("127.0.0.1:0").unwrap();
        listener.set_nonblocking(true).unwrap();
        let addr = listener.local_addr().unwrap();
        let data = Arc::new(Mutex::new(data));
        let etag = Arc::new(Mutex::new("\"v1\"".to_string()));
        let no_range = Arc::new(AtomicBool::new(false));
        let fail_range = Arc::new(AtomicBool::new(false));
        let stall_after = Arc::new(AtomicI64::new(0));
        let speed = Arc::new(AtomicI64::new(0));
        let hits = Arc::new(AtomicUsize::new(0));
        let stop = Arc::new(AtomicBool::new(false));

        let (d, e, n, f, s, sp, st, ht) = (
            data.clone(),
            etag.clone(),
            no_range.clone(),
            fail_range.clone(),
            stall_after.clone(),
            speed.clone(),
            stop.clone(),
            hits.clone(),
        );
        let handle = thread::spawn(move || {
            while !st.load(Ordering::SeqCst) {
                match listener.accept() {
                    Ok((stream, _)) => {
                        let _ = stream.set_nonblocking(false);
                        let (d, e, n, f, s, sp, ht) =
                            (d.clone(), e.clone(), n.clone(), f.clone(), s.clone(), sp.clone(), ht.clone());
                        thread::spawn(move || {
                            let _ = handle_conn(stream, d, e, n, f, s, sp, ht);
                        });
                    }
                    Err(ref err) if err.kind() == std::io::ErrorKind::WouldBlock => {
                        thread::sleep(Duration::from_millis(10));
                    }
                    Err(_) => break,
                }
            }
        });
        TestServer { addr, data, etag, no_range, fail_range, stall_after, speed, hits, stop, handle: Some(handle) }
    }

    fn url(&self) -> String {
        format!("http://{}/file.bin", self.addr)
    }
    fn set_data(&self, d: Vec<u8>, etag: &str) {
        *self.data.lock().unwrap() = d;
        *self.etag.lock().unwrap() = etag.to_string();
    }
    fn set_no_range(&self, v: bool) {
        self.no_range.store(v, Ordering::SeqCst);
    }
    fn set_fail_range(&self, v: bool) {
        self.fail_range.store(v, Ordering::SeqCst);
    }
    fn set_stall_after(&self, n: i64) {
        self.stall_after.store(n, Ordering::SeqCst);
    }
    fn set_speed(&self, bps: i64) {
        self.speed.store(bps, Ordering::SeqCst);
    }
    fn hits(&self) -> usize {
        self.hits.load(Ordering::SeqCst)
    }
}

impl Drop for TestServer {
    fn drop(&mut self) {
        self.stop.store(true, Ordering::SeqCst);
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }
    }
}

fn handle_conn(
    mut stream: TcpStream,
    data: Arc<Mutex<Vec<u8>>>,
    etag: Arc<Mutex<String>>,
    no_range: Arc<AtomicBool>,
    fail_range: Arc<AtomicBool>,
    stall_after: Arc<AtomicI64>,
    speed: Arc<AtomicI64>,
    hits: Arc<AtomicUsize>,
) -> std::io::Result<()> {
    hits.fetch_add(1, Ordering::SeqCst);
    let mut reader = BufReader::new(stream.try_clone()?);
    let mut line = String::new();
    reader.read_line(&mut line)?;
    let mut range: Option<(i64, i64)> = None;
    loop {
        let mut h = String::new();
        let n = reader.read_line(&mut h)?;
        if n == 0 || h == "\r\n" || h == "\n" {
            break;
        }
        let lower = h.to_ascii_lowercase();
        if let Some(rest) = lower.strip_prefix("range:") {
            let size = data.lock().unwrap().len() as i64;
            range = parse_range(rest.trim(), size);
        }
    }

    let body = data.lock().unwrap().clone();
    let etag_v = etag.lock().unwrap().clone();
    let size = body.len() as i64;
    let stall = stall_after.load(Ordering::SeqCst);
    let bps = speed.load(Ordering::SeqCst);

    if no_range.load(Ordering::SeqCst) || range.is_none() {
        let head = format!(
            "HTTP/1.1 200 OK\r\nETag: {etag_v}\r\nAccept-Ranges: none\r\nContent-Length: {size}\r\nConnection: close\r\n\r\n"
        );
        stream.write_all(head.as_bytes())?;
        write_body(&mut stream, &body, stall, bps)?;
        return Ok(());
    }

    let (start, end) = range.unwrap();
    if start > end || start >= size {
        stream.write_all(b"HTTP/1.1 416 Range Not Satisfiable\r\nConnection: close\r\n\r\n")?;
        return Ok(());
    }
    let end = end.min(size - 1);
    let report_start = if fail_range.load(Ordering::SeqCst) && end - start + 1 > 1 {
        start + 1
    } else {
        start
    };
    let head = format!(
        "HTTP/1.1 206 Partial Content\r\nETag: {etag_v}\r\nAccept-Ranges: bytes\r\nContent-Range: bytes {report_start}-{end}/{size}\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
        end - start + 1
    );
    stream.write_all(head.as_bytes())?;
    write_body(&mut stream, &body[start as usize..=end as usize], stall, bps)?;
    Ok(())
}

fn write_body(stream: &mut TcpStream, data: &[u8], stall_after: i64, bps: i64) -> std::io::Result<()> {
    let mut written = 0i64;
    for chunk in data.chunks(32 * 1024) {
        if stall_after > 0 && written >= stall_after {
            thread::sleep(Duration::from_secs(10));
            return Ok(());
        }
        stream.write_all(chunk)?;
        stream.flush()?;
        written += chunk.len() as i64;
        if bps > 0 {
            let secs = chunk.len() as f64 / bps as f64;
            thread::sleep(Duration::from_secs_f64(secs));
        }
    }
    Ok(())
}

fn parse_range(h: &str, size: i64) -> Option<(i64, i64)> {
    let spec = h.strip_prefix("bytes=")?;
    let dash = spec.find('-')?;
    let left = spec[..dash].trim();
    let right = spec[dash + 1..].trim();
    let (start, end) = if left.is_empty() {
        let n: i64 = right.parse().ok()?;
        ((size - n).max(0), size - 1)
    } else {
        let s: i64 = left.parse().ok()?;
        let e = if right.is_empty() { size - 1 } else { right.parse().ok()? };
        (s, e)
    };
    if start < 0 {
        return Some((0, end.min(size - 1)));
    }
    Some((start, end.min(size - 1)))
}

// ---------------- 测试辅助 ----------------

fn make_data(n: usize, seed: u8) -> Vec<u8> {
    (0..n).map(|i| ((i as u64 * 31 + seed as u64 * 7) & 0xff) as u8).collect()
}

fn unique_dir(tag: &str) -> PathBuf {
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    let d = std::env::temp_dir().join(format!("dlcore-{tag}-{nanos}"));
    std::fs::create_dir_all(&d).unwrap();
    d
}

fn test_config(dir: &PathBuf) -> Config {
    let mut c = Config::default();
    c.initial_threads = 1;
    c.max_threads = 4;
    c.min_part_size = 256 << 10;
    c.temp_dir = dir.join("temp");
    c.idle_timeout = Duration::from_secs(3);
    c.max_retries = 2;
    c.retry_delay = Duration::from_millis(50);
    c.adaptive_threads = false; // 测试默认固定并发，便于断言
    c
}

fn read_file(path: &str) -> Vec<u8> {
    std::fs::read(path).unwrap()
}

// ---------------- 测试 ----------------

#[test]
fn download_multithread_matches_source() {
    let data = make_data(4 << 20, 1);
    let srv = TestServer::new(data.clone());
    let dir = unique_dir("multi");
    let target = dir.join("out.bin");
    let engine = Engine::new(test_config(&dir));

    let res = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap();
    assert!(res.range_ok);
    assert!(res.parts >= 2, "期望动态分段 >1 段，实际 {}", res.parts);
    assert_eq!(read_file(&res.path), data);
}

#[test]
fn no_range_falls_back_to_single_stream() {
    let data = make_data(1 << 20, 2);
    let srv = TestServer::new(data.clone());
    srv.set_no_range(true);
    let dir = unique_dir("norange");
    let target = dir.join("out.bin");
    let engine = Engine::new(test_config(&dir));

    let res = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap();
    assert!(!res.range_ok);
    assert_eq!(read_file(&res.path), data);
}

#[test]
fn range_mismatch_is_detected() {
    let data = make_data(1 << 20, 3);
    let srv = TestServer::new(data);
    srv.set_fail_range(true);
    let dir = unique_dir("mismatch");
    let target = dir.join("out.bin");
    let mut cfg = test_config(&dir);
    cfg.initial_threads = 1;
    cfg.max_threads = 1;
    cfg.max_retries = 1;
    let engine = Engine::new(cfg);

    let err = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap_err();
    assert_eq!(err.kind, ErrorKind::Fatal);
}

#[test]
fn idle_timeout_triggers_error() {
    let data = make_data(2 << 20, 4);
    let srv = TestServer::new(data);
    srv.set_stall_after(64 << 10);
    let dir = unique_dir("idle");
    let target = dir.join("out.bin");
    let mut cfg = test_config(&dir);
    cfg.initial_threads = 1;
    cfg.max_threads = 1;
    cfg.idle_timeout = Duration::from_millis(300);
    cfg.max_retries = 1;
    let engine = Engine::new(cfg);

    let err = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap_err();
    assert_eq!(err.kind, ErrorKind::Fatal);
}

#[test]
fn cancel_then_resume() {
    let data = make_data(4 << 20, 5);
    let srv = TestServer::new(data.clone());
    srv.set_speed(2 << 20); // 2 MiB/s，保证能中途取消
    let dir = unique_dir("resume");
    let target = dir.join("out.bin");
    let mut cfg = test_config(&dir);
    cfg.max_retries = 0;
    cfg.idle_timeout = Duration::from_secs(5);
    let engine = Engine::new(cfg);

    let cancel = Arc::new(AtomicBool::new(false));
    let c2 = cancel.clone();
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(400));
        c2.store(true, Ordering::SeqCst);
    });

    let err = engine
        .download(
            Request {
                url: srv.url(),
                target_file: Some(target.display().to_string()),
                cancel: Some(cancel),
                ..Default::default()
            },
            Callbacks::default(),
        )
        .unwrap_err();
    assert_eq!(err.kind, ErrorKind::Canceled);
    assert!(!target.exists(), "取消后不应存在最终文件");

    // 恢复：全速重下剩余部分
    srv.set_speed(0);
    let res = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap();
    assert_eq!(read_file(&res.path), data);
}

#[test]
fn file_changed_on_resume_starts_fresh() {
    let data1 = make_data(2 << 20, 6);
    let data2 = make_data(2 << 20, 7);
    let srv = TestServer::new(data1);
    srv.set_speed(1 << 20);
    let dir = unique_dir("changed");
    let target = dir.join("out.bin");
    let mut cfg = test_config(&dir);
    cfg.max_retries = 0;
    cfg.idle_timeout = Duration::from_secs(5);
    let engine = Engine::new(cfg);

    let cancel = Arc::new(AtomicBool::new(false));
    let c2 = cancel.clone();
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(300));
        c2.store(true, Ordering::SeqCst);
    });
    let err = engine
        .download(
            Request {
                url: srv.url(),
                target_file: Some(target.display().to_string()),
                cancel: Some(cancel),
                ..Default::default()
            },
            Callbacks::default(),
        )
        .unwrap_err();
    assert_eq!(err.kind, ErrorKind::Canceled);

    // 服务器上的文件变了
    srv.set_data(data2.clone(), "\"v2\"");
    srv.set_speed(0);

    let res = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap();
    // 必须是"新文件"，绝不能新旧内容拼在一起
    assert_eq!(read_file(&res.path), data2);
}

#[test]
fn pause_and_resume_completes_correctly() {
    let data = make_data(4 << 20, 8);
    let srv = TestServer::new(data.clone());
    srv.set_speed(2 << 20); // 慢一点，便于中途暂停
    let dir = unique_dir("pause");
    let target = dir.join("out.bin");
    let mut cfg = test_config(&dir);
    cfg.idle_timeout = Duration::from_secs(5);
    let engine = Engine::new(cfg);

    let pause = Arc::new(AtomicBool::new(false));
    let p = pause.clone();
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(300));
        p.store(true, Ordering::SeqCst); // 暂停
        thread::sleep(Duration::from_millis(500));
        p.store(false, Ordering::SeqCst); // 恢复
    });

    let res = engine
        .download(
            Request {
                url: srv.url(),
                target_file: Some(target.display().to_string()),
                pause: Some(pause),
                ..Default::default()
            },
            Callbacks::default(),
        )
        .unwrap();
    assert_eq!(read_file(&res.path), data);
}

#[test]
fn adaptive_threads_grow() {
    let data = make_data(6 << 20, 10);
    let srv = TestServer::new(data.clone());
    srv.set_speed(1 << 20); // 每连接 1 MB/s，慢一点给自适应时间
    let dir = unique_dir("adaptive");
    let target = dir.join("out.bin");
    let mut cfg = test_config(&dir);
    cfg.initial_threads = 1;
    cfg.max_threads = 8;
    cfg.adaptive_threads = true;
    let engine = Engine::new(cfg);

    let t0 = std::time::Instant::now();
    let res = engine
        .download(
            Request { url: srv.url(), target_file: Some(target.display().to_string()), ..Default::default() },
            Callbacks::default(),
        )
        .unwrap();
    let dt = t0.elapsed();
    assert_eq!(read_file(&res.path), data);
    assert!(res.parts >= 2, "自适应应当会加人，实际分段 {}，用时 {:?}", res.parts, dt);
}

#[test]
fn mirrors_are_used() {
    let data = make_data(2 << 20, 11);
    let slow = TestServer::new(data.clone());
    slow.set_speed(1 << 20); // 主源限速
    let fast = TestServer::new(data.clone());
    let dir = unique_dir("mirror");
    let target = dir.join("out.bin");
    let cfg = test_config(&dir);
    let engine = Engine::new(cfg);

    let res = engine
        .download(
            Request {
                url: slow.url(),
                target_file: Some(target.display().to_string()),
                mirrors: vec![fast.url()],
                ..Default::default()
            },
            Callbacks::default(),
        )
        .unwrap();
    assert_eq!(read_file(&res.path), data);
    assert!(fast.hits() > 0, "镜像应当被用到");
}
